src/docGenerator.ts

```diff
@@ -0,0 +1,146 @@
+import fs from "fs"
+import { JSONSchema4 } from "json-schema"
+import { flatMap, flatMapDeep, partition, toPairs } from "lodash"
+import fetch from "node-fetch"
+import { promisify } from "util"
+import { defaultEngine } from "./eslintDefaultOptions"
+import { DescriptionEntry } from "./model/Description"
+import {
+  Category,
+  fromEslintCategoryToCategory,
+  fromEslintCategoryToLevel,
+  Level,
+  patternIdToCodacy,
+  Patterns,
+  PatternsEntry,
+  PatternsParameter
+} from "./model/Patterns"
+import { toolName, toolVersion } from "./toolMetadata"
+
+const writeFile = promisify(fs.writeFile)
+
+export function generatePatterns(): Patterns {
+  const rules = defaultEngine.getRules()
+  const patterns = Array.from(rules.entries()).map(
+    ([patternId, ruleModule]) => {
+      const meta = ruleModule && ruleModule.meta ? ruleModule.meta : undefined
+      const eslintCategory = meta && meta.docs ? meta.docs.category : undefined
+      const level: Level = fromEslintCategoryToLevel(eslintCategory)
+      const category: Category = patternId.includes("security")
+        ? "Security"
+        : fromEslintCategoryToCategory(eslintCategory)
+      const parameters =
+        meta && meta.schema
+          ? fromEslintSchemaToParameters(meta.schema)
+          : undefined
+      return new PatternsEntry(
+        patternIdToCodacy(patternId),
+        level,
+        category,
+        parameters && parameters.length > 0 ? parameters : undefined
+      )
+    }
+  )
+  const entries = patterns.filter(x => x !== null) as PatternsEntry[]
+  return new Patterns(toolName, toolVersion, entries)
+}
+
+export function generateDescription(): DescriptionEntry[] {
+  const rules = defaultEngine.getRules()
+  const descriptionEntries = Array.from(rules.entries()).map(
+    ([patternId, ruleModule]) => {
+      const eslintDescription =
+        ruleModule && ruleModule.meta && ruleModule.meta.docs
+          ? ruleModule.meta.docs.description
+          : undefined
+      const capitalize = (s: string) => s.charAt(0).toUpperCase() + s.slice(1)
+      const description = eslintDescription ? capitalize(eslintDescription) : ""
+      const title = patternId
+        .split("/")
+        .map(s =>
+          capitalize(s)
+            .split("-")
+            .join(" ")
+        )
+        .join(": ")
+      const timeToFix = 5
+      return new DescriptionEntry(
+        patternIdToCodacy(patternId),
+        title,
+        description,
+        timeToFix
+      )
+    }
+  )
+  return descriptionEntries.filter(x => x !== null) as DescriptionEntry[]
+}
+
+type WithDefault = { default: any }
+
+function anyOfToArray(schema: JSONSchema4) {
+  return schema.anyOf ? schema.anyOf : [schema]
+}
+
+function fromEslintSchemaToParameters(
+  schema: JSONSchema4 | JSONSchema4[]
```

Comment:
just `JSONSchema4[]` please

Reply:
Eslint can manage vanilla `JSONSchema4`s and arrays of `JSONSchema4`. And they have different structures when you use the array form or the non array one.
[Source](https://eslint.org/docs/developer-guide/working-with-rules#options-schemas)